src/execution.rs

```diff
@@ -114,6 +122,115 @@ pub trait Report {
     fn start(&mut self, task_name: String);
     fn report(&mut self, result: &CompletedTask);
     fn done(&mut self);
+
+    fn stage(&mut self, full_name: &[String], stage_rep: StageReport) {
+        let mut full_name: Vec<String> = Vec::from(full_name);
+        full_name.push(stage_rep.stage_name);
+        let completed_task = CompletedTask {
+            full_name,
+            duration: stage_rep.duration,
+            stdout: Vec::new(),
+            stderr: Vec::new(),
+            status: Status::from(stage_rep.status),
+        };
+        self.report(&completed_task);
+    }
+}
+
+pub struct StageReportSender {
+    sender: pipe::Sender,
+    // Since we define the stages to be linear, we just need to
+    // keep one timestamp to report a stage's duration.
+    started_at: SystemTime,
+}
+
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub enum StageStatus {
+    Success,
+    Failure(i32),
+    Skipped(String),
+}
+
+impl From<StageStatus> for Status {
+    fn from(s: StageStatus) -> Self {
+        match s {
+            StageStatus::Success => Status::Success,
+            StageStatus::Failure(code) => Status::Failure(code),
+            StageStatus::Skipped(reason) => Status::Skipped(reason),
+        }
+    }
+}
+
+#[derive(PartialEq, Debug, Serialize, Deserialize)]
+pub struct StageReport {
+    stage_name: String,
+    status: StageStatus,
+    duration: Duration,
+}
+
+impl StageReportSender {
+    pub fn report_stage_status<N: ToString>(&mut self, stage_name: N, status: StageStatus) {
+        let stage_name = stage_name.to_string();
+        let end = std::time::SystemTime::now();
```

Comment:
Why do we use SystemTime and not Instant? Unlike SystemTime, Instant is monotonic.

Reply:
Because I didn't know instant existed. :)

Reply:
will change